consensus/src/quorum_store/proof_coordinator.rs

```diff
@@ -67,40 +74,78 @@ impl IncrementalProofState {
             return Err(SignedDigestError::DuplicatedSignature);
         }
 
-        self.aggregated_signature
-            .insert(signed_digest.signer(), signed_digest.signature());
+        match validator_verifier.get_voting_power(&signed_digest.signer()) {
+            Some(voting_power) => {
+                let signer = signed_digest.signer();
+                if self
+                    .aggregated_signature
+                    .insert(signer, signed_digest.signature())
+                    .is_none()
+                {
+                    self.aggregated_voting_power += voting_power as u128;
+                } else {
+                    error!(
+                        "Author already in aggregated_signatures right after rechecking: {}",
+                        signer
+                    );
+                }
+            },
+            None => {
+                error!(
+                    "Received signature from author not in validator set: {}",
+                    signed_digest.signer()
+                );
+                return Err(SignedDigestError::InvalidAuthor);
+            },
+        }
+
         Ok(())
     }
 
     fn ready(&self, validator_verifier: &ValidatorVerifier, my_peer_id: PeerId) -> bool {
-        self.aggregated_signature.contains_key(&my_peer_id)
-            && validator_verifier
-                .check_voting_power(self.aggregated_signature.keys())
-                .is_ok()
+        if self.aggregated_signature.contains_key(&my_peer_id)
+            && self.aggregated_voting_power >= validator_verifier.quorum_voting_power()
+        {
+            let recheck = validator_verifier.check_voting_power(self.aggregated_signature.keys());
+            if recheck.is_err() {
```

Comment:
Should we panic in this case?

Reply:
what is the point in the panic? node can continue operating fine, will collect a few more votes and pass.

we will have alert on error! to actually fix whathever is causing the issue.

that is the main distinction between error!() and panic!(), both should never happen and require code fix, but panic should be used when node doesn't know how to proceed

Reply:
If this is an invariant we are trying to maintain then it's better to panic! right? In this case, something has gone really wrong - may be node's internal state is corrupt and it shouldn't proceed further.

Reply:
I think it is fine to alert and continue, but if you feel strongly I can panic too.

Reply:
IMO it makes sense to panic here but will leave it to you.